### PR TITLE
refactor(algebra): removing unnecessary axioms from group, ring etc.

### PR DESCRIPTION
**Pull request by ChrisHughes24**
This PR minimizes the sets of axioms for various algebraic structures. The removed axioms are

remove `mul_one` from `comm_monoid` and `group`

remove `right_distrib` and `mul_zero` from `comm_semiring`

remove `add_comm` from `ring`

remove `right_distrib` from `comm_ring`

remove `one_mul`, and `inv_mul_cancel` from `division_ring`

## Issues

The changes cause some errors. If, for example, the class `linear_ordered_field` is changed to extend `linear_ordered_comm_ring α, field α` instead of `linear_ordered_comm_ring α, division_ring α`, then there is an error about the types of the two `left_distribs` not matching. Ideally it would extend field rather than `division_ring`, since the `right_distrib` axiom of `division_ring` is unnecessary. I could solve this by simply writing down the axioms explicitly, but I'm not sure if there is some deeper problem that might cause trouble later on.

## Minor points

`one_mul` is not provable constructively in a division ring any more, so I used classical logic.

There is some slight ugly nonuniformity. The axioms for a group already have `a⁻¹ * a = 1` as the inverse axiom, whereas the axioms for a field have `a * a⁻¹ = 1`. This means that in a division ring, `mul_one` is an axiom, but not `one_mul`, whereas in a group `one_mul` is an axiom but not `mul_one`

### CODE REVIEW

File: library/init/algebra/field.lean [hunk 1]
```diff
@@ -16,9 +16,10 @@ set_option default_priority 100
 
 set_option old_structure_cmd true
 
-class division_ring (α : Type u) extends ring α, has_inv α, zero_ne_one_class α :=
+class division_ring (α : Type u) extends
+  add_group α, semigroup α, has_one α, distrib α, has_inv α, zero_ne_one_class α :=
```
Comment (digama0) [inline library/init/algebra/field.lean]: Why isn't a division ring a ring anymore? I don't see the proof below this

Comment (ChrisHughes24) [reply library/init/algebra/field.lean]: There is a proof. `division_ring` doesn't require `one_mul`, but `ring` does. The proof of `one_mul` in a division ring is not entirely trivial, so the proof is some way down.

File: library/init/data/nat/lemmas.lean [hunk 2]
```diff
@@ -119,17 +114,13 @@ instance : comm_semiring nat :=
  add_assoc      := nat.add_assoc,
  zero           := nat.zero,
  zero_add       := nat.zero_add,
- add_zero       := nat.add_zero,
```
Comment (digama0) [inline library/init/data/nat/lemmas.lean]: I think you can delete theorems `nat.add_zero` and similar that were removed from the structures now.

**Comment by digama0** [comment]
Oops, I deleted my comment and apparently your response as well.

File: library/init/data/nat/lemmas.lean (hunk 2 above)
Comment (ChrisHughes24) [reply library/init/data/nat/lemmas.lean]: I already deleted the ones that weren't used in the proofs of other properties.

**Comment by ChrisHughes24** [comment]
@digama0 I don't think this is ready to merge yet because of the issue with `linear_ordered_field`. It should extend `field` rather than `division_ring`, but doing that causes an error.

**Comment by digama0** [comment]
Why don't the types of the two left_distribs match? I would think that either one is a field and the other is a proven property, or they have the same field.

As for one_mul vs mul_one, how about switching the field axiom to `a⁻¹ * a = 1`? It's commutative so it shouldn't matter which we pick, and this makes it more consistent with groups.

**Comment by ChrisHughes24** [comment]
I'm not sure why the types of the two `left_distrib`s don't match. They look like they should reduce to the same thing to me, though they do have different instance paths.

**Comment by ChrisHughes24** [comment]
The downside to changing the axiom for a field to `a⁻¹ * a = 1` is that it will break more stuff, since previously it was the other way, but I guess this isn't really a big deal.

**Comment by digama0** [comment]
his is going to break a lot of stuff anyway. May as well pull the band-aid off and clean things up as much as you can while you have the opportunity.

**Comment by digama0** [comment]
I think the left_distrib problem might be a lean bug. Maybe we should wait until someone has a chance to go into the code and fix it

**Comment by ChrisHughes24** [comment]
One last issue I forgot to mention is that this PR does add a bunch of new instances that aren't really necessary. Now that `group` extends `has_one` but not `monoid`, we have a `group.to_has_one` instance that we didn't have before. Is this a problem?

**Comment by digama0** [comment]
Those instances already existed, with those same names no less. You get a `B.to_A` instance by default when you say `class B extends A`.

**Comment by ChrisHughes24** [comment]
`group.to_has_one` didn't I don't think, because it didn't extend `has_one`, it extended `monoid`

**Comment by digama0** [comment]
Oh you mean transitive instances. That might be a performance problem because lean doesn't handle diamonds well, we'll see.

**Comment by ChrisHughes24** [comment]
If only I could just do `attribute [-instance]` at the end of the file.

**Comment by digama0** [comment]
maybe make an issue :D

**Comment by digama0** [comment]
Could you make a MWE for the left_distrib issue? I'm having trouble reproducing it.

**Comment by ChrisHughes24** [comment]
[MWE](https://gist.github.com/ChrisHughes24/a1f25f9d9e933838ea41585811236212)

**Comment by fpvandoorn** [comment]
> One last issue I forgot to mention is that this PR does add a bunch of new instances that aren't really necessary. Now that `group` extends `has_one` but not `monoid`, we have a `group.to_has_one` instance that we didn't have before. Is this a problem?

A workaround for this is to use 
```lean
structure group ...

attribute [class] group
```
I'm not sure if this has other unintended consequences. I can think of two:
* The projections will have the wrong arguments explicit. This should be fine, since we don't use the `group.foo` projections directly anyway.
* The constructor `group.mk` might be different, but I don't think this will actually be the case with `old_structure_cmd`.

**Comment by ChrisHughes24** [comment]
@fpvandoorn Another downside of the `attribute [class]` approach is that the instances `group.to_has_inv` and the like also have the wrong arguments explicit. I can redefine these as `group.to_has_inv'` etc, but this has caused a problem with a `simp [add_comm]` failing to simplify in a situation where `rw add_comm` solves the goal. 

The `simp` failure fails to prove a goal that looks like `a * b + b * b = b * b + a * b`. The two `a * b` have different paths for `has_mul` and making them the same means that the simplifier succeeds. I know that with `simp [add_comm]` the simplifier puts things in some normalized order, does this ordering depend on the instance path?

**Comment by gebner** [comment]
I'm gonna merge this today.  Please complain if I missed any big issues in the discussion above.

**Comment by gebner** [comment]
BTW, I've reverted this for now: #178.  Let's revisit this in Lean >=3.9.

**Pull request merged**